File: src/main/java/imaxct/dao/impl/BaseDao.java
package imaxct.dao.impl;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.query.Query;

import javax.annotation.Resource;
import java.io.Serializable;
import java.util.List;

/**
 * Created by imaxct on 17-4-10.
 */
public class BaseDao<T> {
    protected static Logger logger = LogManager.getLogger(BaseDao.class);

    @Resource(name = "sessionFactory")
    private SessionFactory sessionFactory;

    protected Session currentSession(){
        return sessionFactory.getCurrentSession();
    }

    public boolean create(T t) {
        Session session = currentSession();
        try {
            session.persist(t);
            return true;
        } catch (Exception e) {
            logger.error("create", e);
            return false;
        }
    }

    public boolean delete(T t) {
        Session session = currentSession();
        try {
            session.delete(t);
            return true;
        } catch (Exception e) {
            session.getTransaction().rollback();
            logger.error("delete", e);
            return false;
        }
    }

    public boolean update(T t) {
        Session session = currentSession();
        try {
            session.update(t);
            return true;
        } catch (Exception e) {
            session.getTransaction().rollback();
            logger.error("update", e);
            return false;
        }
    }

    public boolean saveOrUpdate(T t) {
        Session session = currentSession();
        try {
            session.saveOrUpdate(t);
            return true;
        } catch (Exception e) {
            session.getTransaction().rollback();
            logger.error("saveOrUpdate", e);
            return false;
        }
    }

    public T uniqueResult(String hql, Object... objects) {
        Session session = currentSession();
        Query query = session.createQuery(hql);
        if (objects != null && objects.length > 0) {
            for (int i = 0; i < objects.length; ++i)
                query.setParameter(i, objects[i]);
        }
        return (T) query.uniqueResult();
    }

    public T find(Class<T> clazz, Serializable id) {
        return currentSession().get(clazz, id);
    }

    public List<T> list(String hql, Object... objects) {
        Session session = currentSession();
        Query query = session.createQuery(hql);
        if (objects != null && objects.length > 0) {
            for (int i = 0; i < objects.length; ++i) {
                query.setParameter(i, objects[i]);
            }
        }

        return query.list();
    }

    public List<T> execute(String hql, Object... objects) {
        Session session = currentSession();
        Query query = session.createQuery(hql);
        if (objects != null && objects.length > 0) {
            for (int i = 0; i < objects.length; ++i) {
                query.setParameter(i, objects[i]);
            }
        }
        return query.list();
    }


    public List<T> listPage(int firstPage, int maxResult, String hql, Object... objects) {
        Session session = currentSession();
        Query query = session.createQuery(hql);
        query.setFirstResult(firstPage).setMaxResults(maxResult);
        if (objects != null && objects.length > 0) {
            for (int i = 0; i < objects.length; ++i) {
                query.setParameter(i, objects[i]);
            }
        }
        return query.list();
    }

    public boolean batchUpdate(Object... objects) {
        if (objects != null && objects.length > 0) {
            Session session = currentSession();
            for (Object object : objects) {
                session.update(object);
            }
            return true;
        }
        return false;
    }

    public boolean batchInsert(Object... objects) {
        if (objects != null && objects.length > 0) {
            Session session = currentSession();
            for (Object object : objects) {
                session.persist(object);
            }
            return true;
        }
        return false;
    }
}

File: src/main/java/imaxct/dao/impl/BaseDao.kt
package imaxct.dao.impl

import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.hibernate.Session
import org.hibernate.SessionFactory

import javax.annotation.Resource
import java.io.Serializable

open class BaseDao<T> {

    @Resource(name = "sessionFactory")
    private val sessionFactory: SessionFactory? = null

    private fun currentSession(): Session = sessionFactory!!.currentSession

    fun create(t: T): Boolean {
        val session = currentSession()
        return try {
            session.persist(t)
            true
        } catch (e: Exception) {
            session.transaction.rollback()
            logger.error("create", e)
            false
        }

    }

    fun delete(t: T): Boolean {
        val session = currentSession()
        return try {
            session.delete(t)
            true
        } catch (e: Exception) {
            session.transaction.rollback()
            logger.error("delete", e)
            false
        }

    }

    fun update(t: T): Boolean {
        val session = currentSession()
        return try {
            session.update(t)
            true
        } catch (e: Exception) {
            session.transaction.rollback()
            logger.error("update", e)
            false
        }

    }

    fun saveOrUpdate(t: T): Boolean {
        val session = currentSession()
        return try {
            session.saveOrUpdate(t)
            true
        } catch (e: Exception) {
            session.transaction.rollback()
            logger.error("saveOrUpdate", e)
            false
        }

    }

    fun uniqueResult(hql: String, vararg objects: Any): T {
        val session = currentSession()
        val query = session.createQuery(hql)
        if (objects.isNotEmpty()) {
            for (i in objects.indices)
                query.setParameter(i, objects[i])
        }
        return query.uniqueResult() as T
    }

    fun find(clazz: Class<T>, id: Serializable): T {
        return currentSession().get(clazz, id)
    }

    fun list(hql: String, vararg objects: Any): List<T> {
        val session = currentSession()
        val query = session.createQuery(hql)
        if (objects.isNotEmpty()) {
            for (i in objects.indices) {
                query.setParameter(i, objects[i])
            }
        }

        return query.list() as List<T>
    }

    fun execute(hql: String, vararg objects: Any): List<T> {
        val session = currentSession()
        val query = session.createQuery(hql)
        if (objects.isNotEmpty()) {
            for (i in objects.indices) {
                query.setParameter(i, objects[i])
            }
        }
        return query.list() as List<T>
    }


    fun listPage(firstPage: Int, maxResult: Int, hql: String, vararg objects: Any): List<T> {
        val session = currentSession()
        val query = session.createQuery(hql)
        query.setFirstResult(firstPage).maxResults = maxResult
        if (objects.isNotEmpty()) {
            for (i in objects.indices) {
                query.setParameter(i, objects[i])
            }
        }
        return query.list() as List<T>
    }

    fun batchUpdate(list: List<T>): Boolean {
        if (list.isNotEmpty()) {
            val session = currentSession()
            for ((cnt, obj) in list.withIndex()) {
                session.update(obj)
                if (cnt % 50 == 0){
                    session.flush()
                    session.clear()
                }
            }
            return true
        }
        return false
    }

    fun batchInsert(list: List<T>): Boolean {
        if (list.isNotEmpty()) {
            val session = currentSession()
            for ((cnt, obj) in list.withIndex()) {
                session.persist(obj)
                if (cnt % 50 == 0) {
                    session.flush()
                    session.clear()
                }
            }
            return true
        }
        return false
    }

    fun query(hql: String, vararg objects: Any): Int{
        val session = currentSession()
        val query = session.createQuery(hql)
        for ((i, x) in objects.withIndex()) query.setParameter(i, x)
        return query.executeUpdate()
    }

    companion object {
        protected var logger: Logger = LogManager.getLogger(BaseDao::class.java)
    }
}
